sdk/device_update/src/client.rs

```diff
@@ -0,0 +1,209 @@
+
+/// Client for Device Update operations - import, list and delete updates
+///
+/// # Example
+///
+/// ```no_run
+/// use azure_device_update::DeviceUpdateClient;
+/// use azure_identity::token_credentials::DefaultAzureCredential;
+/// let creds = DefaultAzureCredential::default();
+/// let client = DeviceUpdateClient::new("contoso.api.adu.microsoft.com", &creds).unwrap();
+/// ```
+#[derive(Debug)]
+pub struct DeviceUpdateClient<'a, T> {
+    pub(crate) device_update_url: Url,
+    pub(crate) endpoint: String,
+    pub(crate) token_credential: &'a T,
+    pub(crate) token: Option<TokenResponse>,
+}
+
+impl<'a, T: TokenCredential> DeviceUpdateClient<'a, T> {
+    /// Creates a new `DeviceUpdateClient`.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use azure_device_update::DeviceUpdateClient;
+    /// use azure_identity::token_credentials::DefaultAzureCredential;
+    /// let creds = DefaultAzureCredential::default();
+    /// let client = DeviceUpdateClient::new("contoso.api.adu.microsoft.com", &creds).unwrap();
+    /// ```
+    pub fn new(device_update_url: &str, token_credential: &'a T) -> Result<Self, Error> {
+        let device_update_url = Url::parse(device_update_url)?;
+        let endpoint = extract_endpoint(&device_update_url)?;
+        let client = DeviceUpdateClient {
+            device_update_url,
+            endpoint,
+            token_credential,
+            token: None,
+        };
+        Ok(client)
+    }
+
+    pub(crate) async fn refresh_token(&mut self) -> Result<(), Error> {
+        if matches!(&self.token, Some(token) if token.expires_on > chrono::Utc::now()) {
+            // Token is valid, return it.
+            return Ok(());
+        }
+
+        let token = self
+            .token_credential
+            .get_token(&self.endpoint)
+            .await
+            .map_err(|_| Error::Authorization)?;
+        self.token = Some(token);
+        Ok(())
+    }
+
+    pub(crate) async fn get_authed(&mut self, uri: String) -> Result<String, Error> {
+        self.refresh_token().await?;
+
+        let resp = reqwest::Client::new()
+            .get(&uri)
+            .bearer_auth(self.token.as_ref().unwrap().token.secret())
+            .send()
+            .await;
+        let resp = match resp {
+            Ok(r) => r,
+            Err(_e) => {
+                return Err(Error::InvalidOperationPath());
+            }
+        };
+        let body = resp.text().await.unwrap();
+        Ok(body)
+    }
+
+    pub(crate) async fn _put_authed(&mut self, uri: String, body: String) -> Result<String, Error> {
+        self.refresh_token().await?;
+
+        let resp = reqwest::Client::new()
+            .put(&uri)
+            .bearer_auth(self.token.as_ref().unwrap().token.secret())
+            .header("Content-Type", "application/json")
+            .body(body)
+            .send()
+            .await
+            .unwrap();
+        let body = resp.text().await?;
```

Comment:
We'll want to use the pipeline architecture we're using in `azure_data_ cosmos`.